library/Booster/Syntax/Json/Internalise.hs

```diff
@@ -269,41 +311,111 @@ internaliseTerm ::
+        assocs <- gets (Map.assocs . fst)
+        let sccs =
+                stronglyConnComp [(v, v, Set.toList (freeVariables t)) | (v, t) <- assocs]
+            cycleNodes = Set.fromList [v | CyclicSCC (v : _vs) <- sccs]
+        if null cycleNodes
+            then pure ()
+            else do
+                modify $ \(m, eqs) ->
+                    ( m `Map.withoutKeys` cycleNodes
+                    , eqs <> (map (uncurry mkEq) $ Map.assocs $ m `Map.restrictKeys` cycleNodes)
+                    )
+                breakCycles
+
+mkEq :: Internal.Variable -> Internal.Term -> Internal.Predicate
+mkEq x t = Internal.Predicate $ case sortOfTerm t of
+    Internal.SortInt -> Internal.EqualsInt (Internal.Var x) t
+    Internal.SortBool -> Internal.EqualsBool (Internal.Var x) t
+    otherSort -> Internal.EqualsK (Internal.KSeq otherSort (Internal.Var x)) (Internal.KSeq otherSort t)
+
+internalisePred ::
     Flag "alias" ->
     Flag "subsorts" ->
     Maybe [Syntax.Id] ->
     KoreDefinition ->
     Syntax.KorePattern ->
-    Except PatternError InternalisedPredicate
-internalisePredicate allowAlias checkSubsorts sortVars definition@KoreDefinition{sorts} pat = case pat of
+    Except PatternError [InternalisedPredicate]
+internalisePred allowAlias checkSubsorts sortVars definition@KoreDefinition{sorts} p = case p of
     Syntax.KJEVar{} -> term
     Syntax.KJSVar{} -> term
     Syntax.KJApp{} -> term
     Syntax.KJString{} -> term
-    Syntax.KJTop{} -> pure $ IsPredicate $ Internal.Predicate Internal.TrueBool
-    Syntax.KJBottom{} -> notSupported -- TODO should we throw here?
+    Syntax.KJTop{} -> pure [BoolPred $ Internal.Predicate Internal.TrueBool]
+    Syntax.KJBottom{} -> notSupported
     Syntax.KJNot{arg} -> do
         recursion arg >>= \case
-            IsPredicate (Internal.Predicate (Internal.EqualsInt a b)) -> pure $ IsPredicate $ Internal.Predicate $ Internal.NEqualsInt a b
-            IsPredicate (Internal.Predicate (Internal.EqualsK a b)) -> pure $ IsPredicate $ Internal.Predicate $ Internal.NEqualsK a b
-            IsPredicate (Internal.Predicate p) -> pure $ IsPredicate $ Internal.Predicate $ Internal.NotBool p
-            IsSubstitution k v ->
+            [BoolPred (Internal.Predicate (Internal.EqualsInt a b))] ->
+                pure [BoolPred $ Internal.Predicate $ Internal.NEqualsInt a b]
+            [BoolPred (Internal.Predicate (Internal.NEqualsInt a b))] ->
+                pure [BoolPred $ Internal.Predicate $ Internal.EqualsInt a b]
+            [BoolPred (Internal.Predicate (Internal.EqualsK a b))] ->
+                pure [BoolPred $ Internal.Predicate $ Internal.NEqualsK a b]
+            [BoolPred (Internal.Predicate (Internal.NEqualsK a b))] ->
+                pure [BoolPred $ Internal.Predicate $ Internal.EqualsK a b]
+            [BoolPred (Internal.Predicate p')] ->
+                pure [BoolPred $ Internal.Predicate $ Internal.NotBool p']
+            [SubstitutionPred k v] ->
```

Comment:
what about cases where we have more than one `BoolPred` or `Substitution`? i.e. `#Not (... #And ...)` which we have seen previously

Reply:
All input like that currently go into the `_ -> notSupported` case  (creating an `UnsupportedPred`).
We could support `#Not ( ... #Or ... #Or ... )` for cases where all the `...` can be negated easily (the `Equals* <--> NEquals*`). As we discussed the other day this would mean to implement DeMorgan.
I don't think we can support `#Not ( ... #And ...)` unless the `...` are all of `Bool` sort (and likewise this would mean implementing DeMorgan with assumed `notBool` and `orBool` operations).

Reply:
we wouldnt need demorgan... #Not ( X #And Y ) ~> notBool (X' andBool Y') if X, Y are bool predicates